File: src/Phan_02/constructor_destructor.cpp
#include<iostream>
using namespace std;

class Animal {
private:
    string name;
    int age;
public:
    // default constructor
    Animal() {
        name = "Animal";
        age = 0;
    }

    // parameterized contructor
    Animal(string name, int age) {
        this->name = name;
        this->age = age;
    }

    // copy constructor
    Animal(const Animal& other) {
        this->name = other.name;
        this->age = other.age;
    }

    // destructor
    ~Animal() {
        cout << "Animal destructor" << endl;
    }

    void description() {
		cout<< name << " is "<< age << " years old\n"; 
	}
};

int main() {
    // default constructor
    Animal animal1;
    animal1.description();  // Animal is 0 years old

    // parameterized constructor
    Animal animal2("Rocky", 10);
    animal2.description();  // Rocky is 10 years old 

    // copy constructor
    Animal animal3(animal2);
    animal3.description();  // Rocky is 10 years old
    
    // destructor will call before the program finish
    return 0;
}

File: src/Phan_02/constructor_destructor.py
import copy


class Animal:
    # default constructor
    def __init__(self):
        self.name = "Animal"
        self.age = 0

    # parameterized constructor
    def __init__(self, name, age):
        self.name = name
        self.age = age

    # destructor
    def __del__(self):
        print("Animal destructor")

    def description(self):
        print(f"{self.name} is {self.age} years old")


# default constructor
# animal1 = Animal()    # error 

# parameterized constructor
animal2 = Animal("Rocky", 10)
animal2.description()   # Rocky is 10 years old

# copy class by using copy library
clone1 = copy.copy(animal2)  # shallow copy of animal
clone2 = copy.deepcopy(animal2)  # deep copy of animal

clone1.description()    # Rocky is 10 years old
clone2.description()    # Rocky is 10 years old
